types/coin.go

```diff
@@ -101,127 +132,160 @@ func (coins Coins) String() string {
+func (coins Coins) safePlus(coinsB Coins) Coins {
 	sum := ([]Coin)(nil)
 	indexA, indexB := 0, 0
 	lenA, lenB := len(coins), len(coinsB)
+
 	for {
 		if indexA == lenA {
 			if indexB == lenB {
+				// return nil coins if both sets are empty
 				return sum
 			}
-			return append(sum, coinsB[indexB:]...)
+
+			// return set B (excluding zero coins) if set A is empty
+			return append(sum, removeZeroCoins(coinsB[indexB:])...)
 		} else if indexB == lenB {
-			return append(sum, coins[indexA:]...)
+			// return set A (excluding zero coins) if set B is empty
+			return append(sum, removeZeroCoins(coins[indexA:])...)
 		}
+
 		coinA, coinB := coins[indexA], coinsB[indexB]
+
 		switch strings.Compare(coinA.Denom, coinB.Denom) {
-		case -1:
-			if coinA.IsZero() {
-				// ignore 0 sum coin type
-			} else {
+		case -1: // coin A denom < coin B denom
+			if !coinA.IsZero() {
 				sum = append(sum, coinA)
 			}
+
 			indexA++
-		case 0:
-			if coinA.Amount.Add(coinB.Amount).IsZero() {
-				// ignore 0 sum coin type
-			} else {
-				sum = append(sum, coinA.Plus(coinB))
+
+		case 0: // coin A denom == coin B denom
+			res := coinA.Plus(coinB)
+			if !res.IsZero() {
+				sum = append(sum, res)
 			}
+
 			indexA++
 			indexB++
-		case 1:
-			if coinB.IsZero() {
-				// ignore 0 sum coin type
-			} else {
+
+		case 1: // coin A denom > coin B denom
+			if !coinB.IsZero() {
 				sum = append(sum, coinB)
 			}
+
 			indexB++
 		}
 	}
 }
 
-// Negative returns a set of coins with all amount negative
-func (coins Coins) Negative() Coins {
-	res := make([]Coin, 0, len(coins))
-	for _, coin := range coins {
-		res = append(res, Coin{
-			Denom:  coin.Denom,
-			Amount: coin.Amount.Neg(),
-		})
+// Minus subtracts a set of coins from another.
+//
+// e.g.
+// {2A, 3B} - {A} = {A, 3B}
+// {2A} - {0B} = {2A}
+// {A, B} - {A} = {B}
+//
+// CONTRACT: Minus will never return Coins where one Coin has a non-positive
+// amount. In otherwords, IsValid will always return true.
+func (coins Coins) Minus(coinsB Coins) Coins {
+	diff, hasNeg := coins.SafeMinus(coinsB)
+	if hasNeg {
+		panic("negative coin amount")
 	}
-	return res
+
+	return diff
 }
 
-// Minus subtracts a set of coins from another (adds the inverse)
-func (coins Coins) Minus(coinsB Coins) Coins {
-	return coins.Plus(coinsB.Negative())
+// SafeMinus performs the same arithmetic as Minus but returns a boolean if any
+// negative coin amount was returned.
+func (coins Coins) SafeMinus(coinsB Coins) (Coins, bool) {
+	diff := coins.safePlus(coinsB.negative())
+	return diff, !diff.IsNotNegative()
 }
 
-// IsAllGT returns True iff for every denom in coins, the denom is present at a
+// IsAllGT returns true iff for every denom in coins, the denom is present at a
 // greater amount in coinsB.
 func (coins Coins) IsAllGT(coinsB Coins) bool {
-	diff := coins.Minus(coinsB)
-	if len(diff) == 0 {
+	diff, hasNeg := coins.SafeMinus(coinsB)
+	if len(diff) == 0 || hasNeg {
```

Comment:
This is weird semantically.  False isn't safe to return, esp when people might use !A.IsAllGT(B) as a synonym for A.IsAllLTE(B), etc.